edb/edgeql/compiler/stmt.py

```diff
@@ -254,16 +258,105 @@ def compile_GroupQuery(
     return result
 
 
-@dispatch.compile.register(qlast.InsertQuery)
-def compile_InsertQuery(
-        expr: qlast.InsertQuery, *, ctx: context.ContextLevel) -> irast.Set:
+def simple_stmt_eq(lhs: irast.Base, rhs: irast.Base) -> bool:
+    if (
+        isinstance(lhs, irast.BaseConstant)
+        and isinstance(rhs, irast.BaseConstant)
+        and lhs.typeref == rhs.typeref  # XXX?
+        and lhs.value == rhs.value
+    ):
+        return True
+    elif (
+        isinstance(lhs, irast.Parameter)
+        and isinstance(rhs, irast.Parameter)
+        and lhs.name == rhs.name
+    ):
+        return True
+    else:
+        return False
 
-    if ctx.in_conditional is not None:
+
+def handle_conditional_insert(
+        expr: qlast.InsertQuery,
+        rhs: irast.InsertStmt,
+        lhs_set: Union[irast.Expr, irast.Set],
+        *, ctx: context.ContextLevel,
+) -> List[irast.PointerRef]:
+    def error(s: str) -> NoReturn:
         raise errors.QueryError(
-            'INSERT statements cannot be used inside conditional expressions',
+            f'Invalid conditional INSERT statement: {s}',
             context=expr.context,
         )
 
+    schema = ctx.env.schema
+
+    if not isinstance(lhs_set, irast.Set):
+        error("left hand side is not SELECT")
+    lhs_set = irutils.unwrap_set(lhs_set)
+    if not isinstance(lhs_set.expr, irast.SelectStmt):
+        error("left hand side is not SELECT")
+    lhs = lhs_set.expr
+
+    if lhs.result.path_id != rhs.subject.path_id:
+        error("types do not match")
+
+    if lhs.where:
+        filtered_ptrs = inference.cardinality.extract_filters(
+            lhs.result, lhs.where,
+            scope_tree=ctx.path_scope, singletons=(), env=ctx.env, strict=True)
+    else:
+        filtered_ptrs = None
+
+    # TODO: Can we support some >1 cases?
+    if not filtered_ptrs or len(filtered_ptrs) > 1:
+        error("does not contain exactly one FILTER clause")
+        return None
+
+    exclusive_constr: s_constr.Constraint = schema.get('std::exclusive')
+
+    # Every filter must be on a unique pointer
+    # Adapted from _analyze_filter_caluse
+    for ptr, _ in filtered_ptrs:
+        ptr = ptr.get_nearest_non_derived_parent(schema)
+        is_unique = (
+            ptr.is_id_pointer(schema) or
+            any(c.issubclass(schema, exclusive_constr)
+                for c in ptr.get_constraints(schema).objects(schema))
+        )
+        if not is_unique or not ptr.is_property(schema):
+            error("FILTER is not on an exclusive property")
+
+    shape_props = {}
+    for shape_set, _ in rhs.subject.shape:
+        # what's the base_ptr deal here???
```

Comment:
`Foo { some_ptr }` produces a virtual subtype of `Foo` (let's call it `Foo1`), where `Foo1.some_ptr` is a derivative of `Foo.some_ptr`.  `base_ptr` in `rptr` points to the original `Foo.some_ptr` (which would be the actual property in schema).